### PR TITLE
CHAOS-437: fix node failure examples, add new ddmark rule

### PR DESCRIPTION
**Pull request by nathantournant**
## What does this PR do?

- [X] Adds new functionality
- [ ] Alters existing functionality
- [X] Fixes a bug
- [ ] Improves documentation or testing

Please briefly describe your changes as well as the motivation behind them:
- The example around `nodeFailure` can seem confusing; it's currently easy to start a nodeFailure disruption defaulting to the pod level, with a node-oriented selector, that will most likely find no target (or worse, undesired targets).
- This PR adds a related DDMark rule allowing to solve that issue, by checking a given struct subfield value as a trigger to expect other subfield presence or specific value. Read `ddmark/README.md` for more details.

## Code Quality Checklist

- [X] The documentation is up to date.
- [X] My code is sufficiently commented and passes continuous integration checks.
- [X] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md)).

## Testing

- [X] I leveraged continuous integration testing
    - [X] by depending on existing `unit` tests or `end-to-end` tests.
    - [X] by adding new `unit` tests or `end-to-end` tests.
- [X] I manually tested the following steps:
    - play with `examples/node_failure.yaml`: remove the `nodeFailure` and `level: node` fields and validate with `chaosli` to see the difference. Do not forget to build an update chaosli (`make chaosli`) on the branch, and test with the command`./bin/chaosli/chaosli_darwin_arm64 validate --path examples/node_failure.yaml`
    - [X] locally.
    - [ ] as a canary deployment to a cluster.

### CODE REVIEW

File: ddmark/validation.go [hunk 1]
```diff
@@ -241,3 +297,53 @@ func parseIntOrUInt(value reflect.Value) (int, bool) {
 
 	return fieldInt, ok
 }
+
+// checkValueExistOrIsValid checks if a given string marker item name value exist in a unmarshalled struct (converted to a map by structValueToMap)
+// it returns 1 if the value is found and -if applicable- the required value is valid, 0 otherwise
+func checkValueExistsOrIsValid(markerItem string, structMap map[string]interface{}, ruleName string) (int, error) {
```
Comment (luphaz) [inline ddmark/validation.go:303]: Why using an int instead of a bool?
Even if the method is used by incrementing a counter, it feels to me this method should return a boolean instead. 
It's up to the caller to decide to increment a counter, not to this method purpose IMO

File: ddmark/validation.go [hunk 2]
```diff
@@ -241,3 +297,53 @@ func parseIntOrUInt(value reflect.Value) (int, bool) {
...
+	// marketItem can either be a fieldName, or fieldName=fieldValue
+	markerItemArr := strings.Split(markerItem, "=")
```
Comment (luphaz) [inline ddmark/validation.go:305]: You may want to use cut instead:
* https://pkg.go.dev/strings#Cut

```suggestion
	before, after, found := strings.Cut(markerItem, "=")
```

https://go.dev/play/p/py5xzqG_SgL

File: ddmark/validation.go [hunk 3]
```diff
@@ -241,3 +297,53 @@ func parseIntOrUInt(value reflect.Value) (int, bool) {
...
+			case reflect.Int:
+				vInt := v.Convert(tv).Interface().(int)
+				vStr = strconv.Itoa(vInt)
+			case reflect.String:
+				vStr = v.Convert(t).Interface().(string)
+			default:
+				return 0, fmt.Errorf("%v: please do not apply this marker to anything else than int or string. Current type: %v", ruleName, tv.Name())
+			}
+
+			if strings.Compare(markerSubfieldValue, vStr) == 0 {
```
Comment (luphaz) [inline ddmark/validation.go:337]:
```suggestion
			if markerSubfieldValue == vStr {
```

Using string.Compare is only relevant if you use the various output (-1/1/0) here as it's the simple equality it feels more natural to do a natural comparison (case sensitive).

You could add the option to check for equality without case sensitivity (EqualFold)

File: ddmark/validation.go [hunk 4]
```diff
@@ -241,3 +297,53 @@ func parseIntOrUInt(value reflect.Value) (int, bool) {
...
+			return 1, nil
+		}
+	// a value was required => check item has described value
+	case 2:
+		markerSubfieldName, markerSubfieldValue := markerItemArr[0], markerItemArr[1]
+		if structMap[markerSubfieldName] == nil {
+			break
+		}
+
+		if reflect.Indirect(reflect.ValueOf(structMap[markerSubfieldName])).Type().ConvertibleTo(reflect.TypeOf(markerSubfieldValue)) {
```
Comment (luphaz) [inline ddmark/validation.go:320]: There is a lot of nesting here that makes the "golden path" hard to understand.

In general you should favour escaping when it's not the golden path to reduce nesting.

More details:
* https://www.youtube.com/watch?v=CFRhGnuXG-4&ab_channel=CodeAesthetic
* https://drew.thecsillags.com/posts/2014-04-14-refactoring-flow-control-and-conditionals/

Comment (luphaz) [reply ddmark/validation.go:320]: I would even extract this to a method TBH (extract or even compare int or string or something similar)

File: ddmark/validation_integration_test.go [hunk 5]
```diff
@@ -166,70 +166,148 @@ exclusivefieldstest:
...
+		})
 		It("checks out valid all-nil values", func() {
-			linkedfieldsYaml := `
-linkedfieldstest:
+			linkedfieldsvalueYaml := `
+linkedfieldsvaluetest:
   randomintfield: 1
   strfield:
   pstrfield:
   intfield: 0 # is nil
```
Comment (luphaz) [inline ddmark/validation_integration_test.go:200]:
```suggestion
  intfield: 0 # is empty
```

File: ddmark/validation_integration_test.go [hunk 6]
```diff
@@ -166,70 +166,148 @@ exclusivefieldstest:
...
+			linkedfieldsvaluewithtriggerYaml := `
+linkedfieldsvaluewithtriggertest:
+  strfield: aa      # invalid
+  pstrfield: bb     # invalid
+  intfield: 12
+  pintfield: 1
+  aintfield: [1,2]
+`
+			err := validateString(linkedfieldsvaluewithtriggerYaml)
+			Expect(err.Errors).To(HaveLen(0))
```
Comment (luphaz) [inline ddmark/validation_integration_test.go:258]:
```suggestion
			Expect(err.Errors).To(BeEmpty())
```

FYI this can be used also

File: ddmark/validation_integration_test.go [hunk 7]
```diff
@@ -166,70 +166,148 @@ exclusivefieldstest:
...
 		})
 	})
 
+	Context("LinkedFieldsValueWithTrigger Marker", func() {
+		// +ddmark:validation:LinkedFieldsValueWithTrigger={StrField=aaa,IntField=2}
+		// +ddmark:validation:LinkedFieldsValueWithTrigger={PStrField=bbb,PIntField=12,AIntField}
+		It("checks out valid all-incorrect initial values", func() {
+			linkedfieldsvaluewithtriggerYaml := `
+linkedfieldsvaluewithtriggertest:
+  strfield: aa      # invalid
```
Comment (luphaz) [inline ddmark/validation_integration_test.go:251]: I'm sorry it's a little confusing for me here 🤔 

How can we have invalid as a comment and no error?

File: ddmark/validation_integration_test.go [hunk 8]
```diff
@@ -166,70 +166,148 @@ exclusivefieldstest:
...
+			linkedfieldsvaluewithtriggerYaml := `
+linkedfieldsvaluewithtriggertest:
+  strfield: aaa
+  pstrfield: bbb
+  intfield:  # is nil
+  pintfield: # is nil
+  aintfield: # is nil
+`
+			err := validateString(linkedfieldsvaluewithtriggerYaml)
+			Expect(err.Errors).To(HaveLen(2))
```
Comment (luphaz) [inline ddmark/validation_integration_test.go:283]: I would probably check the expected string to confirm validation rules that triggered is the one expected.

One of the simplest way would be:
```suggestion
			Expect(err.Errors).To(HaveLen(2))
			Expect(err.Errors[0]).To(MatchError("what will be in the Error() string"))
			Expect(err.Errors[1]).To(MatchError("what will be in the other Error() string"))
```

File: ddmark/validation_teststruct.go [hunk 9]
```diff
@@ -67,9 +68,20 @@ type ExclusiveFieldsTestStruct struct {
 	CField    int
 }
 
-// +ddmark:validation:LinkedFields={StrField,IntField}
-// +ddmark:validation:LinkedFields={PStrField,PIntField,AIntField}
-type LinkedFieldsTestStruct struct {
+// +ddmark:validation:LinkedFieldsValue={StrField=aaa,IntField}
+// +ddmark:validation:LinkedFieldsValue={PStrField,PIntField,AIntField}
+type LinkedFieldsValueTestStruct struct {
+	RandomIntField int // allows to actually check all-empty structs
```
Comment (luphaz) [inline ddmark/validation_teststruct.go:74]: 🤔 not sure I get it. If we remove this field what happens?

File: ddmark/validation_unit_test.go [hunk 10]
```diff
@@ -233,6 +233,68 @@ var _ = Describe("Validation Rules Cases", func() {
 			fakeObj.Field2 = 0
 			Expect(linked.ApplyRule(ValueOf(fakeObj))).ToNot(BeNil())
 		})
+
+		It("rejects object with one missing field (Field2 is 3, expected value was 2)", func() {
+			fakeObj.Field2 = 3
+			Expect(linked.ApplyRule(ValueOf(fakeObj))).ToNot(BeNil())
```
Comment (luphaz) [inline ddmark/validation_unit_test.go:239]:
```suggestion
			Expect(linked.ApplyRule(ValueOf(fakeObj))).ToNot(Succeed())
```

It's more readable to use `Succeed` when function returns an error. Then use `To/ToNo` to express what is expected

In general when you expect an error, I tend to prefer checking the EXACT error message:

```suggestion
			Expect(linked.ApplyRule(ValueOf(fakeObj))).To(MatchError("what is inside the error message"))
```

That will check not only error exists AND is the expected one.

File: ddmark/validation_unit_test.go [hunk 11]
```diff
@@ -233,6 +233,68 @@ var _ = Describe("Validation Rules Cases", func() {
...
+		}
+
+		// we expect that; if Field1 value is "aaa" then Field 2 value is 12 and Field3 is not nil
+		arr := []string{"Field1=aaa", "Field2=12", "Field3"}
+		linked := LinkedFieldsValueWithTrigger(arr)
+		var fakeObj dummyStruct
+
+		BeforeEach(func() {
+			i := 3
+			var pi *int = &i
```
Comment (luphaz) [inline ddmark/validation_unit_test.go:257]:
```suggestion
			pi := &i
```

File: ddmark/validation_unit_test.go [hunk 12]
```diff
@@ -233,6 +233,68 @@ var _ = Describe("Validation Rules Cases", func() {
...
+				Field1: "aaa",
+				Field2: 12,
+				Field3: pi,
+			}
+		})
+		It("validates object with all correct fields", func() {
+			Expect(linked.ApplyRule(ValueOf(fakeObj))).To(BeNil())
+		})
+
+		It("validates object with all nil fields", func() {
```
Comment (luphaz) [inline ddmark/validation_unit_test.go:269]:
```suggestion
		It("validates object with all zero fields", func() {
```

**Review by luphaz — COMMENTED** [review]
Nice addition 👏 

It will definitely be helpful to express which field is required or not.

I'm wondering if we can express this in the OpenAPI schema also 🤔

File: ddmark/validation.go (hunk 1 above)
Comment (nathantournant) [reply ddmark/validation.go:303]: agreed

File: ddmark/validation.go (hunk 2 above)
Comment (nathantournant) [reply ddmark/validation.go:305]: I forgot about Cut -- nice. I'll adapt.

Comment (nathantournant) [reply ddmark/validation.go:305]: there's also a point to make about lack of testing for a value field requiring an empty value (`"fieldEmptyValue="`). I'll add test to assert the marker's behaviour.

Comment (nathantournant) [reply ddmark/validation.go:305]: done ! along with an ongoing test rework to adapt to new codebase standards

File: ddmark/validation_unit_test.go (hunk 10 above)
Comment (nathantournant) [reply ddmark/validation_unit_test.go:239]: I agree it adds a greater level of precision and is better in general.
These error messages are customised to every marker definition because of the included field names -- and I don't like the idea of expecting a string hard-coded error, as it adds toil and I think it provides little added value in this situation.
I'll experiment with a error-generation functions in the DDValidationMarker interface, that would allow to standardise the error generation and reduce the amount of hard-coding to do this. It'll allow for a clearer code base and more precise tests.

File: ddmark/validation_teststruct.go (hunk 9 above)
Comment (nathantournant) [reply ddmark/validation_teststruct.go:74]: This might seem a bit intricate; the struct-level marker is only run on a given struct if it has at lease 1 existing subfield. If no subfield is set, the struct is never unmarshalled and just skipped.
All considered this is an expected and desirable behavior, otherwise every possible marked struct subfields would become de-facto Required in the validation of a parent struct.
This field is not here to be tested, but to allow for all-empty struct testing (ie. all 5 testing fields are empty or nil) without seeing the struct be skipped in the integration test.

File: ddmark/validation.go (hunk 4 above)
Comment (nathantournant) [reply ddmark/validation.go:320]: Came from a switch refactoring -- just realised how horrific that code was, adjusting !

File: ddmark/validation_integration_test.go (hunk 7 above)
Comment (nathantournant) [reply ddmark/validation_integration_test.go:251]: phrasing is bad, I'll edit; if the trigger values are not fitting the marker requirement (ie. invalid, incorrect, unfit), the marker validation is valid whatever the other fields' values

File: ddmark/validation_unit_test.go (hunk 10 above)
Comment (nathantournant) [reply ddmark/validation_unit_test.go]: In the end, I compromised with type-interfaced error generation functions for unit tests, and hard-coded strings for integration tests

File: docs/node_disruption.md [hunk 13]
```diff
@@ -5,4 +5,13 @@ The `nodeFailure` field triggers a kernel panic on the node. Because the node wi
 * `/proc/sys/kernel/sysrq` > `/mnt/sysrq`
 * `/proc/sysrq-trigger` > `/mnt/sysrq-trigger`
 
-> :warning:️ Node behavior when using this disruption can differ depending on the cloud provider (node may or may not be replaced, restarted, cordoned, etc.).
\ No newline at end of file
+> :warning:️ Node behavior when using this disruption can differ depending on the cloud provider (node may or may not be replaced, restarted, cordoned, etc.).
+
+## Targeting
+
+For clarity purpose, it is mandatory to explicitely set the disruption's `level` field to either `pod` or `node`.
```
Comment (ptnapoleon) [inline docs/node_disruption.md:12]:
```suggestion
For clarity purpose, it is mandatory to explicitly set the disruption's `level` field to either `pod` or `node`.
```

File: docs/node_disruption.md [hunk 14]
```diff
@@ -5,4 +5,13 @@ The `nodeFailure` field triggers a kernel panic on the node. Because the node wi
...
+
+The `nodeFailure` disruption acts on a node. However, the `level` field needs to be adapted for the selector:
+
+- if `level: node` is set, the selector will target nodes and impact those directly nodes.
```
Comment (ptnapoleon) [inline docs/node_disruption.md:16]:
```suggestion
- if `level: node` is set, the selector will target nodes and impact those nodes directly.
```

File: docs/node_disruption.md [hunk 15]
```diff
@@ -5,4 +5,13 @@ The `nodeFailure` field triggers a kernel panic on the node. Because the node wi
...
+- if `level: pod` is set, the selector will targets pods and impact the nodes that hosts them.
```
Comment (ptnapoleon) [inline docs/node_disruption.md:17]:
```suggestion
- if `level: pod` is set, the selector will targets pods and impact the nodes that host them.
```

File: examples/node_failure.yaml [hunk 16]
```diff
@@ -11,6 +11,7 @@ metadata:
   annotations:
     chaos.datadoghq.com/environment: "lima"
 spec:
+  level: node # selector targets nodes instead of pods. Targeting a pod will impact the node hosting given pod.
   selector:
     app: demo-curl
```
Comment (ptnapoleon) [inline examples/node_failure.yaml:16]: If you're changing the level, you need to change the selector to not be a pod selector anymore

File: examples/node_failure_shutdown.yaml [hunk 17]
```diff
@@ -11,6 +11,7 @@ metadata:
   annotations:
     chaos.datadoghq.com/environment: "lima"
 spec:
+  level: node # selector targets nodes instead of pods. Targeting a pod will impact the node hosting given pod.
```
Comment (ptnapoleon) [inline examples/node_failure_shutdown.yaml:14]: If you're changing the level, you need to change the selector to not be a pod selector anymore

**Review by ptnapoleon — CHANGES_REQUESTED** [review]
The new examples no longer work in our demo environment

File: ddmark/validation.go [hunk 18]
```diff
@@ -151,33 +199,102 @@ func (l LinkedFields) ApplyRule(fieldvalue reflect.Value) error {
...
+
+	if matchCount != 0 && matchCount != len(l) {
+		return l.GenValueCheckError()
+	}
+
+	return nil
+}
+
+func (l LinkedFieldsValueWithTrigger) GenValueCheckError() error {
+	template := "%v: all of the following fields need to be aligned; if the first value is valid / exists, all the following need to either exist or have the indicated value: %v"
```
Comment (ptnapoleon) [inline ddmark/validation.go:274]: Can we possibly extract "the first value" out from `l` here, so it looks like:
```
all of the following fields need to be aligned; if nodeFailure is set, all the following need to either exist or have the indicated value: [level]
```

I think "valid / exists" might be too vague, we can probably assume that if we got here, we're dealing with something valid (or at least something parseable). These messages are primarily for inexperienced users, after all

File: ddmark/README.md [hunk 19]
```diff
@@ -52,10 +52,26 @@ It also allows us to define custom rules to apply to our structures, and focus t
...
+  - without values:
+    -  `// +ddmark:validation:LinkedFieldsValue={<fieldName1>,<fieldName2>,...}`
+    - Applies to: any `<struct>` type
+    - Asserts the fields in the list are either all `nil` or all non-`nil`
+  - with values:
+    - `// +ddmark:validation:LinkedFieldsValue={<fieldName1>=<value>,<fieldName2>,...}`
+    - Applies to: any `<struct>` type -- value checks are only for `string` and `int` subfields
+    - Asserts the fields in the list are either all `nil` (or are *specifically* set to the indicated value) or all non-`nil` (or are ***specifically not*** set to the given value)
+    - Fields with values and without values can be mixed
+- LinkedFieldsValueWithTrigger
```
[the remark this answers is not in the document]
Comment (nathantournant) [reply ddmark/README.md:65]: I am not _happy_ with it, but couldn't really think of better either

File: ddmark/validation.go (hunk 18 above)
Comment (nathantournant) [reply ddmark/validation.go:274]: good point; I'll adapt that now

Comment (nathantournant) [reply ddmark/validation.go]: done !

**Review by ptnapoleon — APPROVED** [review]

File: ddmark/validation_teststruct.go
Comment (luphaz) [inline ddmark/validation_teststruct.go:1]: if this file is only for test suffix it with `_test.go` so it's not available in the built package.

File: docs/node_disruption.md [hunk 20]
```diff
@@ -5,4 +5,13 @@ The `nodeFailure` field triggers a kernel panic on the node. Because the node wi
...
\ No newline at end of file
+> :warning:️ Node behavior when using this disruption can differ depending on the cloud provider (node may or may not be replaced, restarted, cordoned, etc.).
+
+## Targeting
+
+For clarity purpose, it is mandatory to explicitly set the disruption's `level` field to either `pod` or `node`.
+
+The `nodeFailure` disruption acts on a node. However, the `level` field needs to be adapted for the selector:
+
+- if `level: node` is set, the selector will target nodes and impact those nodes directly.
```
Comment (luphaz) [inline docs/node_disruption.md:16]: Hence level is somehow misleading and should be part of selector potentially.

If we revamp the spec we should probably have something like:

```yaml
  selector:
    labels:
      test: value
    annotations:
      test: value
    level: node|pod
```

**Review by luphaz — APPROVED** [review]
